File: src/app/android/app/src/main/java/ch/yarx/jasscardeye/SessionRecorder.kt
package ch.yarx.jasscardeye

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaFormat
import android.media.MediaMuxer
import android.os.ParcelFileDescriptor
import android.os.SystemClock
import android.util.Log
import android.view.Surface
import java.io.OutputStream
import java.util.Locale
import java.util.TreeMap

// Port of src/app/ios/Sources/SessionRecorder.swift.

/**
 * Records a whole counting session as a square video.
 *
 * What it writes is not the camera stream but the **analysed square, frame by frame as the model received
 * it** - same crop, including the frames dropped while an inference was running. Replayed later it follows
 * the run rather than the camera, which is what makes it usable as emulator input and as the basis of a
 * test case. It is not a perfect copy: encoding costs time inside the frame loop, so a recorded session runs
 * at a slightly lower analysis rate than the same session unrecorded.
 *
 * H.264 in MP4 through `MediaCodec`, written straight into Movies/JassCardEye as a pending file that only
 * becomes visible when the session closes (see [PendingMedia]) - so a half-written file never shows up where
 * the captures are, and a long session is not written twice.
 *
 * With the video it writes the **recognition log**, a CSV of the same name in Documents/JassCardEye -
 * MediaStore keeps it out of Movies: one row per frame in the video, saying what the model made of it, row n for
 * frame n. With it goes [sessionInfo], the session info ([SessionInfo]) as JSON: device, app, model and the settings the
 * pile counted with. The encoder may drop a frame it was given, so a row is written only
 * when a frame comes out of the encoder, matched to its analysis by the presentation time [RecordingSurface]
 * stamped on it.
 *
 * When the session ends, the three are packed into one ZIP in Documents/JassCardEye ([SessionArchive]) and the
 * loose files are discarded before anyone sees them, so one file per session is what gets copied off the phone.
 */
class SessionRecorder(private val context: Context, val fileName: String, private val sessionInfo: String? = null) {

    /** How a recording ended. A failure is reported rather than swallowed. */
    sealed interface Outcome {
        /** Packed into the ZIP of this name; [problem] says why packing failed - the files are then published on their own. */
        data class Saved(val name: String, val frames: Int, val problem: String? = null) : Outcome
        data object NothingRecorded : Outcome
        data class Failed(val reason: String) : Outcome
    }

    /** What the model made of one recorded frame: one row of the recognition log, with the same columns as on iOS. */
    data class FrameResult(val detection: Detection? = null, val committed: Boolean = false) {
        /** The row for frame [index] of the video, presented [timeUs] after the first. The box is normalised to the square, origin top-left. */
        fun row(index: Int, timeUs: Long): String {
            val head = "$index,${"%.3f".format(Locale.US, timeUs / 1000.0)},"
            val flag = if (committed) 1 else 0
            val detection = detection ?: return "$head,,,,,,$flag\n"
            val numbers = with(detection) { listOf(confidence, box.left, box.top, box.width, box.height) }
                .joinToString(",") { "%.4f".format(Locale.US, it) }
            return "$head${detection.label},$numbers,$flag\n"
        }

        companion object {
            /** The columns, described under "Session recordings" in context/architecture/data-pipeline.md. */
            const val HEADER = "frame,t_ms,label,confidence,x,y,w,h,committed\n"
        }
    }

    private var renderer: RecordingSurface? = null
    private var csvMedia: PendingMedia? = null
    private var infoMedia: PendingMedia? = null
    private var csv: OutputStream? = null
    private var startedAt: Long? = null
    private var lastSubmitted = -1L
    private var lastWritten = -1L
    private var firstWritten: Long? = null
    private val pending = TreeMap<Long, FrameResult>()
    private val lock = Any()
    private var codec: MediaCodec? = null
    private var surface: Surface? = null
    private var muxer: MediaMuxer? = null
    private var track = -1
    private var muxing = false
    private var side = 0
    private var frameCount = 0
    private var closed = false
    private var failure: String? = null
    private var media: PendingMedia? = null
    private var descriptor: ParcelFileDescriptor? = null
    private val info = MediaCodec.BufferInfo()

    /** Appends the analysed square of this frame. Does nothing once the recording is closed or has failed. */
    fun append(square: Bitmap, result: FrameResult = FrameResult()): Unit = synchronized(lock) {
        // A recording that has already broken must not keep paying for itself inside the inference loop.
        if (closed || failure != null) return
        if (square.width < 16) return
        if (codec == null && !start(minOf(square.width, MAX_SIDE))) return
        val renderer = renderer ?: return
        try {
            drain(endOfStream = false)
            // Bound the queue when an encoder stalls. No submission means no CSV row.
            if (pending.size >= 120) return
            val now = SystemClock.elapsedRealtimeNanos() / 1000
            if (startedAt == null) startedAt = now
            // At least a millisecond apart, so every frame can be found again by its time alone - the Dataset Tool
            // seeks to a tenth of a millisecond before it.
            val time = maxOf(now - checkNotNull(startedAt), lastSubmitted + MIN_SPACING_US)
            pending[time] = result
            renderer.draw(square, time)
            lastSubmitted = time
            drain(endOfStream = false)
        } catch (error: Exception) {
            fail("Aufnahme abgebrochen: ${describe(error)}")
        }
    }

    /** Closes the file and says what became of it. Safe to call more than once. */
    fun finish(): Outcome = synchronized(lock) {
        if (closed) return Outcome.NothingRecorded
        closed = true
        failure?.let { return Outcome.Failed(it) }
        val codec = codec
        if (codec == null || lastSubmitted < 0) {
            release()
            media?.discard(); csvMedia?.discard(); infoMedia?.discard()
            return Outcome.NothingRecorded
        }
        return try {
            codec.signalEndOfInputStream()
            drain(endOfStream = true)
            check(frameCount > 0) { "Der Encoder hat kein Frame geschrieben" }
            csv?.flush(); csv?.close(); csv = null
            // stop writes the MP4 index; a failure here must not publish a broken pair.
            if (muxing) { muxer?.stop(); muxing = false }
            release()
            // One file per session to copy off the phone. A recording that cannot be packed is still a recording:
            // its files are published on their own, and the note says why.
            val problem = pack()
            if (problem == null) {
                Outcome.Saved("$baseName.zip", frameCount)
            } else {
                checkNotNull(csvMedia).commit()
                infoMedia?.commit()
                checkNotNull(media).commit()
                Outcome.Saved(fileName, frameCount, "nicht gepackt: $problem")
            }
        } catch (error: Exception) {
            release()
            media?.discard(); csvMedia?.discard(); infoMedia?.discard()
            Outcome.Failed("Aufnahme nicht abgeschlossen: ${describe(error)}")
        }
    }

    private val baseName: String get() = fileName.substringBeforeLast('.')

    /**
     * Copies the still pending video, log and info into one ZIP in Documents/JassCardEye, publishes it and discards
     * the three. Returns null when that worked, and why not otherwise - with the three left pending.
     */
    private fun pack(): String? {
        val parts = listOfNotNull(
            media?.let { fileName to it },
            csvMedia?.let { "$baseName.csv" to it },
            infoMedia?.let { "$baseName.json" to it },
        )
        var archive: PendingMedia? = null
        return try {
            val entry = PendingMedia.insert(context, PendingMedia.Kind.SESSION_ARCHIVE, "$baseName.zip")
            archive = entry
            val out = context.contentResolver.openOutputStream(entry.uri) ?: error("ZIP konnte nicht geöffnet werden")
            SessionArchive.write(out, baseName, parts.map { (name, part) ->
                name to { context.contentResolver.openInputStream(part.uri) ?: error("$name ist nicht lesbar") }
            })
            entry.commit()
            parts.forEach { (_, part) -> part.discard() }
            null
        } catch (error: Exception) {
            archive?.discard()
            describe(error)
        }
    }

    /** Builds the encoder once the square's size is known - it depends on the lens. */
    private fun start(requested: Int): Boolean {
        var created: MediaCodec? = null
        return try {
            created = MediaCodec.createEncoderByType(MediaFormat.MIMETYPE_VIDEO_AVC)
            // CONFIGURE_FLAG_ENCODE is not optional. Without it the codec is set up in the decoder role and refuses
            // every format with an error that names nothing - the emulator's Codec2 encoder and a Qualcomm OMX
            // encoder alike.
            created.configure(format(created, requested), null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
            surface = created.createInputSurface()
            renderer = RecordingSurface(checkNotNull(surface), side)
            created.start()
            codec = created
            val entry = PendingMedia.insert(context, PendingMedia.Kind.VIDEO, fileName)
            media = entry
            csvMedia = PendingMedia.insert(context, PendingMedia.Kind.SESSION_CSV, fileName.substringBeforeLast('.') + ".csv")
            csv = context.contentResolver.openOutputStream(checkNotNull(csvMedia).uri)
                ?: error("Protokoll konnte nicht geöffnet werden")
            checkNotNull(csv).write(FrameResult.HEADER.toByteArray(Charsets.UTF_8))
            sessionInfo?.let { json ->
                val entry = PendingMedia.insert(context, PendingMedia.Kind.SESSION_INFO, fileName.substringBeforeLast('.') + ".json")
                infoMedia = entry
                context.contentResolver.openOutputStream(entry.uri)?.use { it.write(json.toByteArray(Charsets.UTF_8)) }
                    ?: error("Session-Info konnte nicht geschrieben werden")
            }
            val file = context.contentResolver.openFileDescriptor(entry.uri, "rw")
                ?: throw IllegalStateException("Datei konnte nicht geöffnet werden.")
            descriptor = file
            muxer = MediaMuxer(file.fileDescriptor, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
            Log.i("JassCardEye", "Recording ${side}x$side with ${created.name}")
            true
        } catch (error: Exception) {
            // Once it is `codec`, fail() releases it; before that it is still ours to release.
            if (codec == null) runCatching { created?.release() }
            Log.w("JassCardEye", "Encoder setup failed at ${side}x$side", error)
            fail("Aufnahme konnte nicht angelegt werden: ${describe(error)}")
            false
        }
    }

    private fun format(encoder: MediaCodec, requested: Int): MediaFormat {
        val video = encoder.codecInfo.getCapabilitiesForType(MediaFormat.MIMETYPE_VIDEO_AVC).videoCapabilities
            ?: throw IllegalStateException("kein H.264-Encoder")
        // Not every encoder takes every square: a software encoder on a slow processor stops well below 1080 lines,
        // and hardware encoders have alignments of their own. So the square is the largest the encoder says it can
        // take at the frame rate, never larger than what was asked for.
        val align = maxOf(2, video.widthAlignment, video.heightAlignment)
        side = maxOf(16, requested - requested % align)
        while (side > 16 && !video.areSizeAndRateSupported(side, side, FRAME_RATE.toDouble())) side -= align
        return MediaFormat.createVideoFormat(MediaFormat.MIMETYPE_VIDEO_AVC, side, side).apply {
            setInteger(MediaFormat.KEY_COLOR_FORMAT, MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface)
            setInteger(MediaFormat.KEY_BIT_RATE, video.bitrateRange.clamp(BIT_RATE))
            setInteger(MediaFormat.KEY_FRAME_RATE, FRAME_RATE)
            setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, 1)
            setInteger(MediaFormat.KEY_MAX_B_FRAMES, 0)
        }
    }

    /** An exception as a person can read it - a codec exception often comes with an empty message. */
    private fun describe(error: Exception): String = error.message?.takeIf { it.isNotBlank() } ?: error.javaClass.simpleName

    private fun drain(endOfStream: Boolean) {
        val codec = codec ?: return
        val muxer = muxer ?: return
        val deadline = SystemClock.elapsedRealtime() + 10_000
        while (true) {
            check(!endOfStream || SystemClock.elapsedRealtime() < deadline) { "Encoder beendet die Aufnahme nicht" }
            val index = codec.dequeueOutputBuffer(info, if (endOfStream) 10_000 else 0)
            when {
                index == MediaCodec.INFO_TRY_AGAIN_LATER -> if (!endOfStream) return
                index == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED -> {
                    track = muxer.addTrack(codec.outputFormat)
                    muxer.start()
                    muxing = true
                }
                index >= 0 -> {
                    val data = codec.getOutputBuffer(index)
                    if (data != null && info.size > 0 && muxing && (info.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG) == 0) {
                        data.position(info.offset)
                        data.limit(info.offset + info.size)
                        val pts = info.presentationTimeUs
                        check(pts > lastWritten) { "Encoder liefert Frames in anderer Reihenfolge" }
                        val result = pending.remove(pts) ?: error("Encoder-Frame ohne passenden Analyse-Zeitstempel")
                        pending.headMap(pts).clear() // earlier inputs missing from output were dropped
                        if (firstWritten == null) firstWritten = pts
                        info.presentationTimeUs = pts - checkNotNull(firstWritten)
                        muxer.writeSampleData(track, data, info)
                        checkNotNull(csv).write(result.row(frameCount, info.presentationTimeUs).toByteArray(Charsets.UTF_8))
                        lastWritten = pts
                        frameCount++
                    }
                    codec.releaseOutputBuffer(index, false)
                    if ((info.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) return
                }
            }
        }
    }

    /** Gives up on this recording and takes the half-written file with it. Call with the lock held. */
    private fun fail(reason: String) {
        Log.w("JassCardEye", reason)
        failure = reason
        release()
        media?.discard(); csvMedia?.discard(); infoMedia?.discard()
    }

    private fun release() {
        runCatching { renderer?.close() }; renderer = null
        runCatching { csv?.close() }; csv = null
        pending.clear()
        runCatching { codec?.stop() }
        runCatching { codec?.release() }
        runCatching { surface?.release() }
        if (muxing) runCatching { muxer?.stop() }
        runCatching { muxer?.release() }
        // After the muxer, which writes the file's index on stop and needs the descriptor open for it.
        runCatching { descriptor?.close() }
        descriptor = null
        codec = null
        surface = null
        muxer = null
        muxing = false
    }

    companion object {
        private const val FRAME_RATE = 30
        private const val BIT_RATE = 6_000_000
        /** A recording is evidence, not footage: 1080 px is plenty and keeps the encoder within budget. */
        private const val MAX_SIDE = 1080
        private const val MIN_SPACING_US = 1_000L
    }
}

File: src/app/ios/Sources/SessionRecorder.swift
import AVFoundation
import CoreImage
import CoreVideo
import Foundation

/// Records a whole counting session as a square video.
///
/// What it writes is not the camera stream but the **analysed square, frame by frame as the model
/// received it** - same crop, including the frames dropped while an inference was running. Replayed
/// later it therefore follows the run rather than the camera, which is what makes it usable as
/// simulator input and as the basis of a test case. It is not a perfect copy: a frame the encoder
/// is not ready for is skipped, and encoding costs time inside the frame loop, so a recorded
/// session runs at a slightly lower analysis rate than the same session unrecorded.
///
/// Next to the video it writes the **recognition log**, a CSV of the same name: one row
/// per frame in the video, saying what the model made of it, row n for frame n. A row is written only
/// once the writer has taken the frame, with that frame's presentation time, so a skipped frame leaves
/// no row behind.
///
/// When the session ends, video, log and session info are packed into one ZIP (`SessionArchive`), so
/// one file per session is what gets copied off the phone.
///
/// Frames arrive on the camera queue and `finish()` is called from the main actor, so the little
/// state there is guarded.
final class SessionRecorder: @unchecked Sendable {

    /// How a recording ended. A failure is reported rather than swallowed: a broken recording and a
    /// successful empty one look identical from outside, and the whole point of recording is to
    /// have the evidence afterwards.
    enum Outcome {
        /// Written and closed, and packed into the ZIP of this name. `problem` says why packing failed -
        /// the files are then kept together in a folder of the session's name.
        case saved(name: String, frames: Int, problem: String?)
        /// Nothing was ever written; no file left behind.
        case nothingRecorded
        /// The recording broke. The partial file has been removed; `reason` is shown to the user.
        case failed(reason: String)
    }

    /// What the model made of one recorded frame: one row of the recognition log, with the same
    /// columns as on Android.
    struct FrameResult: Sendable {
        /// The columns, described under "Session recordings" in context/architecture/data-pipeline.md.
        static let header = "frame,t_ms,label,confidence,x,y,w,h,committed\n"

        var label: String? = nil
        var confidence: Float = 0
        /// Vision's box: normalised to the analysed square, origin bottom-left.
        var box: CGRect = .zero
        var committed: Bool = false

        /// The row for frame `index` of the video, presented at `time`. The box is written with a
        /// top-left origin, as Android has it and as the Dataset Tool draws it.
        func row(index: Int, time: CMTime) -> String {
            let head = "\(index),\(Self.number(time.seconds * 1000, digits: 3)),"
            let flag = committed ? "1" : "0"
            guard let label else { return head + ",,,,,,\(flag)\n" }
            let numbers = [Double(confidence), box.minX, 1 - box.maxY, box.width, box.height]
                .map { Self.number($0, digits: 4) }
            return head + ([label] + numbers + [flag]).joined(separator: ",") + "\n"
        }

        private static func number(_ value: Double, digits: Int) -> String {
            String(format: "%.\(digits)f", locale: Locale(identifier: "en_US_POSIX"), value)
        }
    }

    private var csvURL: URL { url.deletingPathExtension().appendingPathExtension("csv") }
    private var infoURL: URL { url.deletingPathExtension().appendingPathExtension("json") }
    /// The session info (`SessionInfo`) as JSON, written beside the log when the recording starts.
    private let info: Data?
    private var csv: FileHandle?
    private var lastTime = CMTime.invalid
    private let url: URL
    private let context = CIContext(options: [.useSoftwareRenderer: false])

    private let lock = NSLock()
    private var writer: AVAssetWriter?
    private var input: AVAssetWriterInput?
    private var adaptor: AVAssetWriterInputPixelBufferAdaptor?
    private var startedAt: CFAbsoluteTime?
    private var closed = false
    private var frameCount = 0

    /// Set once the recording has broken, and never cleared. It stops the retry loop and it is what
    /// `finish()` reports.
    private var failure: String?

    init(url: URL, info: Data? = nil) {
        self.url = url
        self.info = info
    }

    /// Appends the analysed square of this frame. Called on the camera queue; does nothing once the
    /// recording is closed or has failed, and skips a frame while the encoder is not ready - a
    /// dropped frame is far better than a stalled capture.
    func append(_ buffer: CVPixelBuffer, region: CGRect, result: FrameResult = .init()) {
        lock.lock()
        defer { lock.unlock() }
        // A recording that has already broken must not keep paying for itself. This runs on the
        // camera queue, which is the inference loop: without this guard a writer that fails in
        // second 20 of a count would go on rendering a CIImage and taking a pool buffer per frame
        // for the remaining 40 seconds, for nothing.
        guard !closed, failure == nil else { return }

        let image = CIImage(cvPixelBuffer: buffer)
        let extent = image.extent
        guard extent.width > 0, extent.height > 0 else { return }

        // The same rectangle FrameCapture cuts - literally, so the recording and a still taken from
        // the same session can never show different squares.
        let rect = FrameGeometry.crop(region, in: extent)
        guard rect.width >= 16, rect.height >= 16 else { return }

        // start() records its own failure, which the guard above then catches on the next frame.
        if writer == nil, !start(side: Int(min(rect.width, rect.height))) { return }
        guard let writer, let input, let adaptor else { return }

        // The encoder reports a file it can no longer write - a full disk is the common one - in
        // its status rather than by refusing the next append.
        guard writer.status != .failed else {
            fail(writer.error.map { "Aufnahme abgebrochen: \($0.localizedDescription)" }
                 ?? "Aufnahme abgebrochen.", cancelling: writer)
            return
        }

        guard input.isReadyForMoreMediaData, let pool = adaptor.pixelBufferPool else { return }

        var target: CVPixelBuffer?
        guard CVPixelBufferPoolCreatePixelBuffer(nil, pool, &target) == kCVReturnSuccess,
              let target else { return }

        // Moved to the origin so the crop fills the frame; rendering a CIImage into a pixel buffer
        // keeps CoreImage's bottom-left origin on both sides, so the picture stays the right way up.
        let moved = image.transformed(by: .init(translationX: -rect.minX, y: -rect.minY))
        context.render(moved, to: target)

        let now = CFAbsoluteTimeGetCurrent()
        if startedAt == nil { startedAt = now }
        let elapsed = now - (startedAt ?? now)
        // The return value is the only report of a rejected frame - a timestamp collision, say,
        // since timescale 600 quantises to 1.67 ms. Counting the frame regardless would leave a
        // number that says the recording went fine.
        var presentation = CMTime(seconds: max(0, elapsed), preferredTimescale: 600)
        if lastTime.isValid, presentation <= lastTime {
            presentation = lastTime + CMTime(value: 1, timescale: 600)
        }
        guard adaptor.append(target, withPresentationTime: presentation) else {
            fail(writer.error.map { "Aufnahme abgebrochen: \($0.localizedDescription)" }
                 ?? "Aufnahme abgebrochen - Bild abgewiesen.", cancelling: writer)
            return
        }
        do {
            guard let csv else { throw CocoaError(.fileWriteUnknown) }
            try csv.write(contentsOf: Data(result.row(index: frameCount, time: presentation).utf8))
        } catch {
            fail("Erkennungsprotokoll konnte nicht geschrieben werden: \(error.localizedDescription)", cancelling: writer)
            return
        }
        lastTime = presentation
        frameCount += 1
    }

    /// What `finish()` found when it closed the recording, taken under the lock.
    private struct Closing {
        let writer: AVAssetWriter?
        let input: AVAssetWriterInput?
        let frames: Int
        let failure: String?
    }

    /// Closes the gate and hands back what was there. Separate from `finish()` and deliberately
    /// synchronous: a lock must not be taken in an async function - it is held across a suspension
    /// point the compiler cannot see, and from Swift 6 on it is an error rather than a warning.
    /// Nothing in here awaits, so the whole critical section belongs on this side of the boundary.
    private func close() -> Closing? {
        lock.lock()
        defer { lock.unlock() }
        guard !closed else { return nil }
        closed = true
        do { try csv?.synchronize(); try csv?.close(); csv = nil }
        catch { fail("Erkennungsprotokoll konnte nicht abgeschlossen werden: \(error.localizedDescription)", cancelling: writer) }
        return Closing(writer: writer, input: input, frames: frameCount, failure: failure)
    }

    /// Closes the file and says what became of it. Safe to call more than once.
    func finish() async -> Outcome {
        guard let closing = close() else { return .nothingRecorded }
        let wrote = closing.frames

        // fail() has already cancelled the writer and removed the file.
        if let failure = closing.failure { return .failed(reason: failure) }

        guard let writer = closing.writer, let input = closing.input, wrote > 0 else {
            // Nothing recorded - leave no empty file behind.
            try? FileManager.default.removeItem(at: url)
            try? FileManager.default.removeItem(at: csvURL)
            try? FileManager.default.removeItem(at: infoURL)
            return .nothingRecorded
        }

        input.markAsFinished()
        await writer.finishWriting()
        guard writer.status == .completed else {
            try? FileManager.default.removeItem(at: url)
            try? FileManager.default.removeItem(at: csvURL)
            try? FileManager.default.removeItem(at: infoURL)
            return .failed(reason: writer.error.map {
                "Aufnahme nicht abgeschlossen: \($0.localizedDescription)"
            } ?? "Aufnahme nicht abgeschlossen.")
        }
        // One file per session to copy off the phone. A recording that cannot be packed is still a
        // recording: its files stay together in a folder, and the note says why.
        let name = url.deletingPathExtension().lastPathComponent
        let files = [url, csvURL, infoURL].filter { FileManager.default.fileExists(atPath: $0.path) }
        do {
            let archive = try SessionArchive.pack(files, name: name)
            return .saved(name: archive.lastPathComponent, frames: wrote, problem: nil)
        } catch {
            return .saved(name: name, frames: wrote, problem: "nicht gepackt: \(error.localizedDescription)")
        }
    }

    /// Gives up on this recording and takes the half-written file with it. Call with the lock held.
    ///
    /// Removing the file matters as much as the flag: `AVAssetWriter(outputURL:fileType:)` throws
    /// when something is already at the URL, so a leftover would turn every later start into a
    /// throw.
    private func fail(_ reason: String, cancelling target: AVAssetWriter?) {
        failure = reason
        if target?.status == .writing { target?.cancelWriting() }
        writer = nil
        input = nil
        adaptor = nil
        try? csv?.close()
        csv = nil
        try? FileManager.default.removeItem(at: csvURL)
        try? FileManager.default.removeItem(at: infoURL)
        try? FileManager.default.removeItem(at: url)
    }

    /// Builds the writer once the square's size is known - it depends on the lens, so it cannot be
    /// decided before the first frame arrives. Returns false and records the reason on failure.
    ///
    /// Every exit here has to clean up. `startWriting()` fails only *after* the writer has created
    /// the file, so a start that gives up without removing it leaves a zero-byte .mov that makes
    /// the next frame's attempt throw "file exists" - and the frame after that, for the rest of the
    /// session, at thirty throwing allocations a second, without anyone being told.
    private func start(side: Int) -> Bool {
        guard !FileManager.default.fileExists(atPath: url.path),
              !FileManager.default.fileExists(atPath: csvURL.path) else {
            failure = "Aufnahme existiert bereits. Bitte erneut starten."
            return false
        }
        // H.264 wants even dimensions.
        let size = max(16, side - side % 2)
        let settings: [String: Any] = [
            AVVideoCodecKey: AVVideoCodecType.h264,
            AVVideoWidthKey: size,
            AVVideoHeightKey: size,
        ]

        let writer: AVAssetWriter
        do {
            writer = try AVAssetWriter(outputURL: url, fileType: .mov)
        } catch {
            fail("Aufnahme konnte nicht angelegt werden: \(error.localizedDescription)",
                 cancelling: nil)
            return false
        }

        let input = AVAssetWriterInput(mediaType: .video, outputSettings: settings)
        input.expectsMediaDataInRealTime = true
        guard writer.canAdd(input) else {
            fail("Aufnahme: Video-Eingang abgelehnt.", cancelling: writer)
            return false
        }
        writer.add(input)

        let adaptor = AVAssetWriterInputPixelBufferAdaptor(
            assetWriterInput: input,
            sourcePixelBufferAttributes: [
                kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
                kCVPixelBufferWidthKey as String: size,
                kCVPixelBufferHeightKey as String: size,
                // Vision needs IOSurface-backed buffers, and a recording is meant to be replayed
                // through the same path - so the frames are made the same way.
                kCVPixelBufferIOSurfacePropertiesKey as String: [:],
            ])

        guard writer.startWriting() else {
            fail(writer.error.map { "Aufnahme konnte nicht gestartet werden: \($0.localizedDescription)" }
                 ?? "Aufnahme konnte nicht gestartet werden.", cancelling: writer)
            return false
        }
        do {
            try Data(FrameResult.header.utf8).write(to: csvURL, options: .withoutOverwriting)
            csv = try FileHandle(forWritingTo: csvURL)
            try csv?.seekToEnd()
            try info?.write(to: infoURL, options: .withoutOverwriting)
        } catch {
            fail("Erkennungsprotokoll konnte nicht angelegt werden: \(error.localizedDescription)", cancelling: writer)
            return false
        }
        writer.startSession(atSourceTime: .zero)

        self.writer = writer
        self.input = input
        self.adaptor = adaptor
        // The first accepted frame starts at zero; a leading gap creates an extra
        // empty sample in AVAssetReader and would offset every CSV index.
        return true
    }
}
